### DISCUSSION TITLE
Problem with understanding part of the decoder logic

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#52] [Archived]
<USER>>>> damiankwasny95<br/> <DATE> [May 28, 2019, 1:29am]<br/> <br/> Hello. I have a trouble understanding a certain line in the decoder<br/> forward method. This one:<br/> <https://github.com/mozilla/TTS/blob/924d6ad4e55e7763e61933bf3542dae1e892c369/layers/tacotron.py#L405-L417><br/> <br/> memory_input, attention_rnn_hidden, decoder_rnn_hiddens, *slash* <br/> current_context_vec, attention, attention_cum = self._init_states(inputs)<br/> while True:<br/> if t > 0:<br/> if memory is None:<br/> new_memory = outputs[-1]<br/> else:<br/> new_memory = memory[t - 1]<br/> # Queuing if memory size defined else use previous prediction only.<br/> if self.memory_size > 0:<br/> memory_input = torch.cat([memory_input[:, self.r curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz self.memory_dim:].clone(), new_memory], dim=-1)<br/> else:<br/> memory_input = new_memory<br/> <br/> So my question is, what does this concatenation do? We have a<br/> `new_memory` with size `(B, r*memory_channels)` and in step `t=1` when<br/> we can enter this part of the condition, the current `memory_input` we<br/> are trying to concatenate with is of shape<br/> `(B, memory_size*memory_channels)`. So in case `memory_size` is as a<br/> default set to `memory_size = r` this line literally concatenates with<br/> an empty tensor. The only situation it actually concatenates with<br/> something is only when `memory_size` is higher then `r`, am I right?<br/> shouldn't the condition be changed to `if self.memory_size > r` or<br/> something like that in this case?<br/> <DATE> [Maybe I am missing something, but I am] stuck on this line for like more<br/> then half an hours already and I cannot see a different solution. I<br/> would be glad for clarification.<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/problem-with-understanding-part-of-the-decoder-logic]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [May 28, 2019, 8:55am]<br/> <br/> if `memory_size` *slash* > 0, then TTS uses a constant queue regardless of `r`.<br/> So if the queue is active, it basically queues the incoming frames.<br/> Otherwise, it uses the previous prediction as `memory_input`.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> damiankwasny95<br/> <DATE> [June 3, 2019, 4:52am]<br/> <br/> Thanks for the reply! I am still not sure if I understand this part<br/> correctly, but it seems that the code changed anyway, so I will try to<br/> dig into once again and comeback here in case I still don't grasp it.<br/> Have a nice day!<br/> <br/> [Archived Post]<br/>
